### PR TITLE
VAAPI interop for Vulkan

### LINKED ISSUE #5613: How to make vaapi zero copy hwdec work with vulkan?
The performance of opengl with the RAVU shader compares to the vulkan is too low. 
### mpv version and platform
```
mpv 0.28.0-360-g013a8f75f3 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.8.100
   libavcodec      58.14.100
   libavformat     58.10.100
   libswscale      5.0.102
   libavfilter     7.12.100
   libswresample   3.0.101
ffmpeg version: N-90264-g80798e3857
```
```
libva info: VA-API version 1.1.0
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_1
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.1 (libva 2.1.0)
vainfo: Driver version: Mesa Gallium driver 18.1.0-devel for AMD Radeon (TM) RX 460 Graphics (POLARIS11 / DRM 3.23.0 / 4.15.7-1-ARCH, LLVM 7.0.0)
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileNone                   :	VAEntrypointVideoProc
```
### Reproduction steps
`--vo=gpu --gpu-context=x11-egl --gpu-api=vulkan --vulkan-queue-count=8 --hwdec=vaapi`
### Expected behavior
Pixel Format = vaapi
### Actual behavior
Pixel Format = yuv420p
### Log file
https://0x0.st/sm3f.txt

### PR DESCRIPTION
**Pull request by philipl**
This change introduces a vulkan interop path for the vaapi hwdec.
The basic principles are mostly the same as for EGL, with the
exported dma_buf being imported by Vukan. The biggest difference
is that we cannot reuse the texture as we do with OpenGL - there's
no way to rebind a VkImage to a different piece of memory, as far
as I can see. So, a new texture is created on each map call.

I did not bother implementing a code path for the old libva API as
I think it's safe to assume any system with a working vulkan driver
will have access to a newer libva.

Note that we are using separate layers for the vaapi surface, just
as is done for EGL. This is because libplacebo doesn't support
multiplane images.

This change does not include format negotiation because no driver
implements the `vk_ext_image_drm_format_modifier` extension that
would be required to do that. In practice, the two formats we care
about (nv12, p010) work correctly, so we are not blocked. A separate
change had to be made in libplacebo to filter out non-fatal validation
errors related to surface sizes due to the lack of format negotiation.

Fixes #5613

### CODE REVIEW

File: video/out/hwdec/hwdec_vaapi.c [hunk 1]
```diff
@@ -156,19 +169,38 @@ static int init(struct ra_hwdec *hw)
...
+            return -1;
+
+        MP_VERBOSE(hw, "using VAAPI EGL interop\n");
+    }
+#endif
+
+#if HAVE_VULKAN
+    if (ra_pl_get(hw->ra)) {
+        const struct pl_gpu *gpu = ra_pl_get(hw->ra);
+        if (!(gpu->import_caps.buf & PL_HANDLE_DMA_BUF)) {
```
Comment (haasn) [inline video/out/hwdec/hwdec_vaapi.c]: You import the dmabuf as a texture, not a buffer, so you should be checking `import_caps.tex`

**Comment by laichiaheng** [comment]
Any progress on this?

**Comment by laichiaheng** [comment]
Would @haasn review it again?

**Comment by philipl** [comment]
With the validation error filtering added to libplacebo, this change is no longer blocked from merging, assuming reviewers are happy.

**Comment by ReekyMarko** [comment]
Sorry for the noise but I'm very eager to try this feature, however I have no idea how to enable it. I cloned the philipl:libplacebo-vaapi branch and built it. Tried with `--hwdec=vaapi --gpu-api=vulkan` and that doesn't get me hwdec. Is there some other parameters I'm missing?

**Comment by barolo** [comment]
@ReekyMarko 

I can't find it in the code but for egl it was "gpu-hwdec-interop=vaapi-egl" logically it should be something like "gpu-hwdec-interop=vaapi-vk" or "gpu-hwdec-interop=vaapi-vulan" now

**Comment by ReekyMarko** [comment]
> @ReekyMarko
> 
> I can't find it in the code but for egl it was "gpu-hwdec-interop=vaapi-egl" logically it should be something like ~"gpu-hwdec-interop=vaapi-vk"~ or "gpu-hwdec-interop=vaapi-vulkan" now
> Edit: It's vaapi-vulkan

Okay thanks. With this parameter, mpv prints `No hwdec backend named 'vaapi-vulkan' found!`. I guess I must have missed something while building. Maybe it's better for me to wait until this is at least merged to master :laughing:

**Comment by philipl** [comment]
I never changed the name - it's still `vaapi-egl`. Ideally you'd probably rename it to `vaapi` but that would break existing configs.

**Comment by laichiaheng** [comment]
@philipl What parameter do we need to add if it is merged in the future?

**Comment by barolo** [comment]
@philipl oh, so interop just picks whichever gpu-api is stated?

**Comment by barolo** [comment]
> > @ReekyMarko
> > I can't find it in the code but for egl it was "gpu-hwdec-interop=vaapi-egl" logically it should be something like ~"gpu-hwdec-interop=vaapi-vk"~ or "gpu-hwdec-interop=vaapi-vulkan" now
> > Edit: It's vaapi-vulkan
> 
> Okay thanks. With this parameter, mpv prints `No hwdec backend named 'vaapi-vulkan' found!`. I guess I must have missed something while building. Maybe it's better for me to wait until this is at least merged to master laughing

@ReekyMarko as stated above you just need vaapi-egl, if it works hwaccel won't error out with "--hwdec=vaapi --gpu-api=vulkan"

**Comment by philipl** [comment]
Interops don't pick, they are picked. The interop mechanism is actually gpu-api neutral. There is only one list. It just happens to be the case that init will fail if the api isn't supported by the interop mechanism. So even before I did this work, `vaapi-egl` would be tried if you leave the interop mechanism as `auto` (which is really the easiest thing to do). That's why there's only one name and why the name hasn't changed.

**Comment by ReekyMarko** [comment]
> > > @ReekyMarko
> > > I can't find it in the code but for egl it was "gpu-hwdec-interop=vaapi-egl" logically it should be something like ~"gpu-hwdec-interop=vaapi-vk"~ or "gpu-hwdec-interop=vaapi-vulkan" now
> > > Edit: It's vaapi-vulkan
> > 
> > 
> > Okay thanks. With this parameter, mpv prints `No hwdec backend named 'vaapi-vulkan' found!`. I guess I must have missed something while building. Maybe it's better for me to wait until this is at least merged to master laughing
> 
> @ReekyMarko as stated above you just need vaapi-egl, if it works hwaccel won't error out with "--hwdec=vaapi --gpu-api=vulkan"

Okay comparing these two commands:
`mpv video.webm --vo=gpu --hwdec=vaapi`
`mpv video.webm --vo=gpu --hwdec=vaapi --gpu-hwdec-interop=vaapi-egl  --gpu-api=vulkan`

I don't get the (hwdec: vaapi) when I press shift+i and I can tell from CPU usage that hwdec really is not working.

**Comment by philipl** [comment]
@ReekyMarko Please attach a full verbose log.

**Comment by barolo** [comment]
@philipl 
Attaching  log:
https://xv.skro.pl/?59c978edcc5df8ff#FFEE1TL/yOByCSh5hv3a5fUxzLXbe97inhpTD4Ug1yY=

There's something about dma_buf
"[vo/gpu/vaapi-egl] VAAPI Vulkan interop requires support for dma_buf import in Vulkan.                                                               
[vo/gpu] Loading failed."

**Comment by philipl** [comment]
@barolo

AMD doesn't support the Vulkan dma_buf extension yet so interop cannot work. Also, you were trying to use vaapi-copy which, by definition, doesn't use interop.

**Comment by barolo** [comment]
@philipl 
checked it with vaapi and errror is the same, which extension is required? "VK_EXT_external_memory_dma_buf" ?
My vkinfo https://lnk.skro.pl/7

**Comment by philipl** [comment]
> @philipl
> checked it with vaapi and errror is the same, which extension is required? "VK_EXT_external_memory_dma_buf" ?
> My vkinfo https://lnk.skro.pl/7

Yes. AMD doesn't support that extension so no Vulkan interop is possible.

**Comment by barolo** [comment]
@philipl 
but since Vulkaninfo states that my device hast this extension available how I'm supposed to understand this?, is it missing somewhere else in the stack?

**Comment by philipl** [comment]
You can see in the log:

`[vo/gpu/vulkan/libplacebo]       tex import caps: 0x0`

That means it doesn't support importing textures at all. It only supports dma_buf import for buffers, which isn't what we need.

**Comment by barolo** [comment]
@philipl 
Thank you for the explanation, much appreciated

**Comment by LaserEyess** [comment]
I've been getting some crashes with [intel-media-driver](https://github.com/intel/media-driver) as my vaapi driver instead of [intel-vaapi-driver](https://github.com/intel/intel-vaapi-driver). Both seem to have the same capabilities according to vainfo, at least in terms of decoding. 

intel-media-driver: https://0x0.st/z2XJ.txt
intel-vaapi-driver: https://0x0.st/z2XO.txt

For any video I've tried, the following command will work with intel-vaapi-driver but fail with intel-media-driver
```
mpv --no-config --gpu-api=vulkan --hwdec=vaapi $file
```

You can install both drivers at the same time and switch between them with the `LIBVA_DRIVER_NAME` variable. No matter what I do I can never get intel-media-driver to be recognized as default so I have to export `LIBVA_DRIVER_NAME=iHD` to get it to work. So all the logs I created using intel-media-driver have that variable exported.

Here's a log of the crash: https://0x0.st/z2Xv.log
Here's a log using vaapi-driver: https://0x0.st/z2Xw.log

And a full backtrace https://0x0.st/z2XY.txt

**Comment by LaserEyess** [comment]
The new commit fixes the crash

**Comment by laichiaheng** [comment]
Is there any plan to merge this into master branch?

**Comment by laichiaheng** [comment]
> it doesn't support importing textures at all. It only supports dma_buf import for buffers

What should we suggest AMD to do with it? They must support it!

**Comment by philipl** [comment]
You can go ask the radv developers to add support for dma_buf texture imports.

**Comment by laichiaheng** [comment]
I just post an e-mail at mesa-users list, not sure if they will see it.

**Comment by soredake** [comment]
@laichiaheng any success with this?

**Comment by laichiaheng** [comment]
@soredake I don't know.

**Comment by DistantThunder** [comment]
I have this on amdvlk with Vega64 :
```

[...]
GPU id       : 0 (AMD RADV/ACO VEGA10 (LLVM 8.0.1))
Device Extensions       count = 73
[...]
        VK_EXT_external_memory_dma_buf      : extension revision  1

```

And thus it does seem to work (confirmed in OSD):
```

mpv --msg-level=vo=v  --hwdec=vaapi-copy  --gpu-api=vulkan  4K_HDR_HEVC_sample.mkv
[...]
[vo/gpu] Probing for best GPU context.
[vo/gpu/vulkan] Initializing GPU context 'waylandvk'
[vo/gpu/vulkan] Creating instance with extensions:
[vo/gpu/vulkan]     VK_KHR_surface
[vo/gpu/vulkan]     VK_KHR_wayland_surface
Xlib:  extension "NV-GLX" missing on display ":0".
WARNING: Experimental compiler backend enabled. Here be dragons! Incorrect rendering, GPU hangs and/or resets are likely
[vo/gpu/vulkan] Initializing GPU context 'x11vk'
[vo/gpu/vulkan] Creating instance with extensions:
[vo/gpu/vulkan]     VK_KHR_surface
[vo/gpu/vulkan]     VK_KHR_xlib_surface
Xlib:  extension "NV-GLX" missing on display ":0".
WARNING: Experimental compiler backend enabled. Here be dragons! Incorrect rendering, GPU hangs and/or resets are likely
[vo/gpu/x11] X11 opening display: :0
[vo/gpu/x11] X11 running at 2560x1440 (":0" => local display)
[vo/gpu/x11] Detected wm supports NetWM.
[vo/gpu/x11] Detected wm supports FULLSCREEN state.
[vo/gpu/x11] Detected wm supports ABOVE state.
[vo/gpu/x11] Detected wm supports BELOW state.
[vo/gpu/x11] Display 0 (DisplayPort-1): [0, 0, 2560, 1440] @ 143.999210 FPS
[vo/gpu/x11] Current display FPS: 143.999210
[vo/gpu/vulkan] Probing for vulkan devices:
[vo/gpu/vulkan]     GPU 0: AMD RADV/ACO VEGA10 (LLVM 8.0.1) (discrete)
[vo/gpu/vulkan] Chose device:
[vo/gpu/vulkan]     Device Name: AMD RADV/ACO VEGA10 (LLVM 8.0.1)
[vo/gpu/vulkan]     Device ID: 1002:687f
[vo/gpu/vulkan]     Driver version: 79695971
[vo/gpu/vulkan]     API version: 1.1.107
[vo/gpu/vulkan] Initializing SPIR-V compiler 'shaderc'
[vo/gpu/vulkan] Queue families supported by device:
[vo/gpu/vulkan]     QF 0: flags 0xf num 1
[vo/gpu/vulkan]     QF 1: flags 0xe num 4
[vo/gpu/vulkan] Using graphics queue (QF 0)
[vo/gpu/vulkan] Using async transfer (QF 1)
[vo/gpu/vulkan] Creating vulkan device with extensions:
[vo/gpu/vulkan]     VK_KHR_swapchain
[vo/gpu/vulkan] Allocating 1048576 memory of type 0x1 (id 0) in heap 0.
[vo/gpu/vulkan] Allocating 1048576 memory of type 0x1 (id 0) in heap 0.
[vo/gpu] Testing FBO format rgba16hf
[vo/gpu] Using FBO format rgba16hf.
[vo/gpu] Assuming 144.000000 FPS for display sync.
Using hardware decoding (vaapi-copy).

```

I have to explicitly select `vaapi-copy` though, "auto" hwdec falls back on CPU decoding.

**Comment by laichiaheng** [comment]
@DistantThunder We need zero copy vaapi.

**Comment by jstkdng** [comment]
@DistantThunder 
vaapi-copy doesnt use interop, so it isn't really hardware accelerated.

**Comment by laichiaheng** [comment]
It's sad to see AMD doesn't support hardware decoding with Vulkan, although they embrace the opensource world.

**Comment by barolo** [comment]
> 
> 
> It's sad to see AMD doesn't support hardware decoding with Vulkan, although they embrace the opensource world.

But they do? it sprang out recently though, via amd amf
https://www.phoronix.com/scan.php?page=news_item&px=FFmpeg-AMD-AMF-Vulkan

**Comment by laichiaheng** [comment]
@barolo Any way to decode video by AMF in mpv?

**Comment by barolo** [comment]
@laichiaheng 
I have no idea actually, it's all very new and became available under Linux in ffmpeg very recently. I've seen windows mpv builds with amf mentions, would have to check it there later

**Comment by philipl** [comment]
There is no support for the AMF decoder API in ffmpeg or mpv, so that's not available and won't be any time soon. I'd love to see a standardised Vulkan API supported by all vendors, but I mostly expect we'll see both nvidia and intel drag their feet on it because they feel their existing solutions are fine. Only AMD has a strong motivation to adopt it.

**Comment by travankor** [comment]
@laichiaheng 

Any benchmarks with this? I just did some crude testing and vulkan+vaapi seems to use the same CPU as opengl+vaapi (older Intel laptop CPU).

**Comment by philipl** [comment]
For the record, it looks like AMD does support the interop, but it has additional requirements around how memory is bound and that requires new functionality in libplacebo. I guess I'll look at it when I can find some time. It looks like it wouldn't require any mpv changes.

**Comment by haasn** [comment]
^ for the record, this is blocked by the fact that the mesa vaapi implementation requires nonzero offsets for planar formats, while the vulkan driver supports zero offsets only.

Upstream is investigating (tm)

**Comment by soredake** [comment]
https://code.videolan.org/videolan/libplacebo/merge_requests/72

**Pull request merged**